### ISSUE TITLE
pretrained models to get the reported results

### ISSUE THREAD

**Issue by kbjpc123** [dwromero/g_selfatt#4]
Hi,

On the CIFAR 10 dataset, I aim to replicate the given results.
However, the training on the CIFAR-10 dataset always diverge even if I disable automatic mixed precision by using cosine learning rate scheduler.
Could you please share your configuration files so that I can replicate the reported results?

Based on the paper, here is my configuration file for training on the CIFAR-10 dataset.

def get_config():
    default_config = dict(
        # --------------------------
        # General parameters
        dataset="CIFAR10",
        # The dataset to be used, e.g., MNIST.
        model="z2sa",
        # The model to be used, e.g., p4sa.
        optimizer="SGD",
        # The optimizer to be used, e.g., Adam.
        optimizer_momentum=0.9,
        # If optimizer == SGD, this specifies the momentum of the SGD.
        device="cuda",
        # The device in which the model will be deployed, e.g., cuda.
        scheduler="linear_warmup_cosine",
        # The lr scheduler to be used, e.g., multistep, constant.
        sched_decay_steps=(400,),
        # If scheduler == multistep, this specifies the steps at which
        # The scheduler should be decreased.
        sched_decay_factor=1.0,
        # The factor with which the lr will be reduced, e.g., 5, 10.
        lr=0.01,
        # The lr to be used, e.g., 0.001.
        norm_type="LayerNorm",
        # The normalization type to be used in the network, e.g., LayerNorm.
        attention_type="Local",
        # The type of self-attention to be used in the network, e.g., Local, Global.
        activation_function="Swish",
        # The activation function used in the network. e.g., ReLU, Swish.
        patch_size=5,
        # If attention_type == Local, the extension of the receptive field on which self-attention is calculated.
        dropout_att=0.1,
        # Specifies a layer-wise dropout factor applied on the computed attention coefficients, e.g., 0.1.
        dropout_values=0.1,
        # Specifies a layer-wise dropout factor applied on the resulting value coefficients from self-att layers, e.g., 0.1.
        whitening_scale=1.0,
        # Specifies a factor with which the current variance initialization is weighted.
        weight_decay=0.0001,
        # Specifies a L2 norm over the magnitude of the weigths in the network, e.g., 1e-4.
        batch_size=24,
        # The batch size to be used, e.g., 64.
        epochs=350,
        # The number of epochs to perform training, e.g., 200.
        seed=0,
        # The seed of the run. e.g., 0.
        comment="",
        # An additional comment to be added to the config.path parameter specifying where
        # the network parameters will be saved / loaded from.
        pretrained=False,
        # Specifies if a pretrained model should be loaded.
        train=True,
        # Specifies if training should be performed.
        augment=False,  # **No augment used in our experiments.**
        path="",
        # This parameter is automatically derived from the other parameters of the run. It specifies
        # the path where the network parameters will be saved / loaded from.
    )
    default_config = ml_collections.ConfigDict(default_config)
    return default_config

Aside from that, could you please answer a couple of questions?
1. Which attention type did you employ for the CIFAR-10 dataset? local or global?
2. There are no pre-trained models that can reproduce the reported results. Where can I find them?

Best,
Byungjin Kim

**Comment by dwromero** [comment]
Hi Byungjin,

Thank you for your question and apologies for my late answer. 

I just checked the project and unfortunately, I did not save the checkpoint of the trained model. I am very sorry for that. 

However, I did verify the command I used to train the model. It is:

`run_experiment.py --dataset CIFAR10 --model mz2sa --norm_type LayerNorm --attention_type Local --activation_function Swish --patch_size 5 --dropout_att 0.1 --dropout_values 0.0 --whitening_scale 1.41421356 --epochs 350 --optimizer SGD --lr 0.01 --optimizer_momentum 0.9 --scheduler linear_warmup_cosine --optimizer_decay_steps 1000 --optimizer_decay_factor 1.0 --weight_decay 0.0001 --batch_size 24 --device cuda --seed 0 --comment`

Please let me know if this helps. 

Here is the log of the model so that you have an idea of how the training of the model should look like: 
```
GPU's available: 2
Let's use 2 GPUs!
Files already downloaded and verified
Files already downloaded and verified
{'dataset': 'CIFAR10', 'model': 'mz2sa', 'optimizer': 'SGD', 'device': 'cuda:0', 'scheduler': 'linear_warmup_cosine', 'optimizer_decay_steps': [1000], 'optimizer_decay_factor': 1.0, 'lr': 0.01, 'optimizer_momentum': 0.9, 'norm_type': 'LayerNorm', 'attention_type': 'Local', 'activation_function': 'Swish', 'patch_size': 5, 'dropout_att': 0.1, 'dropout_values': 0.0, 'whitening_scale': 1.41421356, 'weight_decay': 0.0001, 'batch_size': 24, 'epochs': 350, 'seed': 0, 'comment': '', 'pretrained': False, 'augment': False, 'path': './saved/model_mz2sa_optim_SGD_lr_0.01_scheduler_linear_warmup_cosine_dpatt_0.1_dpvals_0.0_normtype_LayerNorm_wd_0.0001_seed_0/model.pth'}
2020-09-18 17:55:51.627692
Epoch 1/350
------------------------------
Learning Rate: 0.0
------------------------------
[W TensorIterator.cpp:918] Warning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are in channels_last format. (function operator())
train Loss: 4.4601 Acc: 0.1666
2020-09-18 18:17:43.895278
validation Loss: 2.0901 Acc: 0.2223
2020-09-18 18:19:34.213710
Epoch 2/350
------------------------------
Learning Rate: 0.00100048007681229
------------------------------
train Loss: 2.0019 Acc: 0.2686
2020-09-18 18:41:27.984009
validation Loss: 2.1950 Acc: 0.2286
2020-09-18 18:44:41.826347
Epoch 3/350
------------------------------
Learning Rate: 0.00200096015362458
------------------------------
train Loss: 1.7845 Acc: 0.3494
2020-09-18 19:06:34.077478
validation Loss: 2.0566 Acc: 0.2467
2020-09-18 19:10:06.887950
Epoch 4/350
------------------------------
Learning Rate: 0.00300144023043687
------------------------------
train Loss: 1.6522 Acc: 0.3956
2020-09-18 19:34:02.653140
validation Loss: 2.0699 Acc: 0.3008
2020-09-18 19:38:10.794597
Epoch 5/350
------------------------------
Learning Rate: 0.00400192030724916
------------------------------
train Loss: 1.5757 Acc: 0.4251
2020-09-18 20:00:01.315454
validation Loss: 1.7955 Acc: 0.3572
2020-09-18 20:02:47.101516
Epoch 6/350
------------------------------
Learning Rate: 0.00500240038406145
------------------------------
train Loss: 1.5168 Acc: 0.4501
2020-09-18 20:24:38.576314
validation Loss: 1.7866 Acc: 0.3891
2020-09-18 20:25:53.499778
Epoch 7/350
------------------------------
Learning Rate: 0.00600288046087374
------------------------------
train Loss: 1.4706 Acc: 0.4699
2020-09-18 20:47:45.894475
validation Loss: 2.1764 Acc: 0.2841
2020-09-18 20:49:00.343440
Epoch 8/350
------------------------------
Learning Rate: 0.0070033605376860295
------------------------------
train Loss: 1.4488 Acc: 0.4815
2020-09-18 21:10:51.775935
validation Loss: 1.5424 Acc: 0.4455
2020-09-18 21:12:06.649012
Epoch 9/350
------------------------------
Learning Rate: 0.00800384061449832
------------------------------
train Loss: 1.4103 Acc: 0.5001
2020-09-18 21:33:59.245753
validation Loss: 1.6658 Acc: 0.4125
2020-09-18 21:35:14.107047
Epoch 10/350
------------------------------
Learning Rate: 0.009004320691310611
------------------------------
train Loss: 1.3994 Acc: 0.5094
2020-09-18 21:57:06.845392
validation Loss: 1.4751 Acc: 0.4781
2020-09-18 21:58:21.622690
Epoch 11/350
------------------------------
Learning Rate: 0.009999999995080699
------------------------------
train Loss: 1.3705 Acc: 0.5235
2020-09-18 22:20:14.547627
validation Loss: 1.4806 Acc: 0.5039
2020-09-18 22:21:29.478521
Epoch 12/350
------------------------------
Learning Rate: 0.009999784298278028
------------------------------
train Loss: 1.2884 Acc: 0.5555
2020-09-18 22:43:22.791305
validation Loss: 1.5843 Acc: 0.4545
2020-09-18 22:44:37.440576
Epoch 13/350
------------------------------
Learning Rate: 0.009999141326973908
------------------------------
train Loss: 1.2397 Acc: 0.5749
2020-09-18 23:06:29.939221
validation Loss: 1.2987 Acc: 0.5498
2020-09-18 23:07:45.036380
Epoch 14/350
------------------------------
Learning Rate: 0.009998071136115757
------------------------------
train Loss: 1.2056 Acc: 0.5897
2020-09-18 23:29:38.619152
validation Loss: 1.4039 Acc: 0.5303
2020-09-18 23:30:53.299183
Epoch 15/350
------------------------------
Learning Rate: 0.00999657381716057
------------------------------
train Loss: 1.1774 Acc: 0.5996
2020-09-18 23:52:47.233390
validation Loss: 1.4042 Acc: 0.5278
2020-09-18 23:54:02.100846
Epoch 16/350
------------------------------
Learning Rate: 0.009994649498067115
------------------------------
train Loss: 1.1209 Acc: 0.6222
2020-09-19 00:15:54.228572
validation Loss: 1.2696 Acc: 0.5781
2020-09-19 00:17:09.157035
Epoch 17/350
------------------------------
Learning Rate: 0.00999229834328498
------------------------------
train Loss: 1.0856 Acc: 0.6358
2020-09-19 00:39:00.615022
validation Loss: 1.5353 Acc: 0.5060
2020-09-19 00:40:15.331317
Epoch 18/350
------------------------------
Learning Rate: 0.00998952055374053
------------------------------
train Loss: 1.0517 Acc: 0.6477
2020-09-19 01:02:08.561455
validation Loss: 1.7826 Acc: 0.4838
2020-09-19 01:03:23.342028
Epoch 19/350
------------------------------
Learning Rate: 0.009986316366819736
------------------------------
train Loss: 1.0182 Acc: 0.6582
2020-09-19 01:25:17.795362
validation Loss: 1.3919 Acc: 0.5542
2020-09-19 01:26:32.698099
Epoch 20/350
------------------------------
Learning Rate: 0.009982686056347886
------------------------------
train Loss: 0.9888 Acc: 0.6691
2020-09-19 01:48:26.453611
validation Loss: 1.4741 Acc: 0.5420
2020-09-19 01:49:41.413253
Epoch 21/350
------------------------------
Learning Rate: 0.009978629932566178
------------------------------
train Loss: 0.9698 Acc: 0.6761
2020-09-19 02:11:35.597115
validation Loss: 1.0833 Acc: 0.6643
2020-09-19 02:12:50.449228
Epoch 22/350
------------------------------
Learning Rate: 0.009974148342105223
------------------------------
train Loss: 0.9459 Acc: 0.6840
2020-09-19 02:34:43.494143
validation Loss: 1.2836 Acc: 0.5993
2020-09-19 02:35:58.071275
Epoch 23/350
------------------------------
Learning Rate: 0.009969241667955412
------------------------------
train Loss: 0.9227 Acc: 0.6932
2020-09-19 02:57:50.791810
validation Loss: 1.3364 Acc: 0.6016
2020-09-19 02:59:05.771158
Epoch 24/350
------------------------------
Learning Rate: 0.009963910329434177
------------------------------
train Loss: 0.9152 Acc: 0.6982
2020-09-19 03:20:59.783907
validation Loss: 1.4010 Acc: 0.5791
2020-09-19 03:22:14.798708
Epoch 25/350
------------------------------
Learning Rate: 0.009958154782150182
------------------------------
train Loss: 0.8957 Acc: 0.7057
2020-09-19 03:44:09.179257
validation Loss: 1.9385 Acc: 0.4915
2020-09-19 03:45:24.114551
Epoch 26/350
------------------------------
Learning Rate: 0.00995197551796436
------------------------------
train Loss: 0.8696 Acc: 0.7129
2020-09-19 04:07:18.970802
validation Loss: 1.1021 Acc: 0.6602
2020-09-19 04:08:33.956925
Epoch 27/350
------------------------------
Learning Rate: 0.009945373064947902
------------------------------
train Loss: 0.8471 Acc: 0.7207
2020-09-19 04:30:28.569619
validation Loss: 1.1376 Acc: 0.6475
2020-09-19 04:31:43.747611
Epoch 28/350
```

**Comment by kbjpc123** [comment]
Hi,

Thank you for your thoughtful comments.
Unfortunately, even after using the command you provided, I am still experiencing the same issue.
By the way, it appears that your code baseline is not the same as mine.

First, when I attempted to run tests with your command, I received the following errors.

```
python run_experiment.py  --dataset CIFAR10 --model mz2sa --norm_type LayerNorm --attention_type Local --activation_fuction Swish --patch_size 5 --dropout_att 0.1 --dropout_values 0.0 --whitening_scale 1.41421356 --epochs 350 --optimizer SGD --lr 0.01 --optimizer_momentum 0.9 --scheduler linear_warmup_cosine --optimizer_decay_steps 1000 --optimizer_decay_factor 1.0 --weight_decay 0.0001 --batch_size 24 --device cuda --side 0 --comment
FATAL Flags parsing error: Unknown command line flag 'dataset'
Pass --helpshort or --helpfull to see help on flags.
```

Second, the arguments 'optimizer decay step' and 'optimizer decay factor' are missing from the config.py file.
Instead of them, I believe you mean 'sched_decay_steps' and 'sched_decay_factor'.
So, I tried below command.

`python run_experiment.py --config.dataset CIFAR10 --config.model mz2sa --config.norm_type LayerNorm --config.attention_type Local --config.activation_function Swish --config.patch_size 5 --config.dropout_att 0.1 --config.dropout_values 0.0 --config.whitening_scale 1.41421356 --config.epochs 350 --config.optimizer SGD --config.lr 0.01 --config.optimizer_momentum 0.9 --config.scheduler linear_warmup_cosine --config.sched_decay_steps='(1000,)' --config.sched_decay_factor 1.0 --config.weight_decay 0.0001 --config.batch_size 24 --config.device cuda --config.seed 0 --config.comment ''` 

However, I continue to experience NaN loss, as shown below.
```
GPU's available: 2
Let's use 2 GPUs!
Number of parameters: 3001808
Files already downloaded and verified
Files already downloaded and verified
activation_function: Swish
attention_type: Local
augment: false
batch_size: 24
comment: ''
dataset: CIFAR10
device: cuda:0
dropout_att: 0.1
dropout_values: 0.0
epochs: 350
lr: 0.01
model: mz2sa
norm_type: LayerNorm
optimizer: SGD
optimizer_momentum: 0.9
patch_size: 5
path: saved/CIFAR10_augm_False_model_mz2sa_type_Local_patch_5_dpatt_0.1_dpval_0.0_activ_Swish_norm_LayerNorm_white_1.41421356_optim_SGD_momentum_0.9_lr_0.01_bs_24_ep_350_wd_0.0001_seed_0_sched_linear_warmup_cosine.pt
pretrained: false
sched_decay_factor: 1.0
sched_decay_steps: !!python/tuple
- 1000
scheduler: linear_warmup_cosine
seed: 0
train: true
weight_decay: 0.0001
whitening_scale: 1.41421356
2022-02-17 01:25:30.951731
No AMP will be used. Schedulers other than constant make models trained with AMP diverge. Current: linear_warmup_cosine
Epoch 1/350
------------------------------
Learning Rate: 0.0
------------------------------
tensor(44.0584, device='cuda:0', grad_fn=<NllLossBackward>)
[W TensorIterator.cpp:918] Warning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are in channels_last format. (function operator())
tensor(nan, device='cuda:0', grad_fn=<NllLossBackward>)
tensor(nan, device='cuda:0', grad_fn=<NllLossBackward>)
tensor(nan, device='cuda:0', grad_fn=<NllLossBackward>)
tensor(nan, device='cuda:0', grad_fn=<NllLossBackward>)
```

Note that I included debug code to print the loss value, as shown below.
```
diff --git a/trainer.py b/trainer.py
index e06bbc8..b8b3d45 100644
--- a/trainer.py
+++ b/trainer.py
@@ -68,6 +68,7 @@ def train(model, dataloaders, config):

                         outputs = model(inputs)
                         loss = criterion(outputs, labels)
+                        print(loss)
                         if train:
                             loss.backward()
                             optimizer.step()
```

You stated in your paper that you disable automatic mixed precision; but, when I enable it, I get an identical outcome to yours.
Do you recall if you turned on automatic mixed precision?
```
------------------------------
Learning Rate: 0.0
------------------------------
[W TensorIterator.cpp:918] Warning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are in channels_last format. (function operator())
train Loss: 4.7966 Acc: 0.1494
2022-02-17 02:34:42.946061
validation Loss: 2.2419 Acc: 0.2129
2022-02-17 02:36:10.733137
Epoch 2/350
------------------------------
Learning Rate: 0.00100048007681229
------------------------------
```

Do you have any suggestions for resolving this problem?

Best,
Byungjin Kim
